analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/NullPointerDereference.cs

```diff
@@ -41,27 +41,35 @@ public override bool ShouldExecute()
         {
             var walker = new SyntaxKindWalker();
             walker.SafeVisit(Node);
-            return walker.Result;
+            return walker.Result ?? false;
         }
 
         private sealed class SyntaxKindWalker : SafeCSharpSyntaxWalker
         {
-            public bool Result { get; private set; }
+            public bool? Result { get; private set; }
 
             public override void Visit(SyntaxNode node)
             {
-                if (node.IsAnyKind(
-                    SyntaxKind.AwaitExpression,
-                    SyntaxKind.ElementAccessExpression,
-                    SyntaxKind.ForEachStatement,
-                    SyntaxKind.SimpleMemberAccessExpression))
+                if (Result is false)
```

Comment:
We need to stop on `true` as well as `false` - whenever we know the answer
```suggestion
                if (!Result.HasValue)
```

Reply:
If `Result == true` we still need to continue because we may find later an `?.` or `??`. We can only stop processing in the `false` case.